[Feature] Career development form

yonikid15 opened the pull request:
🤖 Resolves #12011 

## 👋 Introduction

- Adds the career development form to the employee profile page
- Uses @esizer BoolCheckIcon component from the [Community interest dialogs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12524), which may need tweaking
## 🕵️ Details

- There's going to be backend and frontend changes to this form in the near future. The figma design for this section has already been updated, so you'll need to view the images below to review everything.

## 🧪 Testing

1. Build app
2. Login as `applicant-employee@test.com`
3. Go to `en/applicant/employee-profile`
4. Ensure career development section matches design and functions correctly

## 📸 Screenshot

![image (2)](https://github.com/user-attachments/assets/ca109f0b-7fa0-4f73-a7cb-c0a07c203f62)

![image (4)](https://github.com/user-attachments/assets/dfdf4fce-4007-499f-aef2-cbbd9b2ed507)

![image (3)](https://github.com/user-attachments/assets/45129731-c3b6-423b-98bd-4e11a8a2f2cb)

brindasasi reviewed (changes requested):
* The work counter is working weird.  It doesn't count when I add the word in the next line. 
![Screenshot 2025-02-06 at 1 37 36 PM](https://github.com/user-attachments/assets/52046202-3b35-444b-b8be-2ac50b6938ce)

* For the larger text it looks like this . Could we wrap it ? 
![Screenshot 2025-02-06 at 1 39 54 PM](https://github.com/user-attachments/assets/eec147ee-2b10-4cd7-a77f-e15dc0ba19e9)


* I'm unable to get the missing information screen shot that you have posted above.  As a new government employee I get these sections as below 

![Screenshot 2025-02-06 at 1 54 16 PM](https://github.com/user-attachments/assets/810985f1-a7fa-4eed-bbc5-7c00ac279cb7)

esizer commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,503 @@
+    );
+  };
+
+  const dataToFormValues = ({
+    organizationTypeInterest,
+    moveInterest,
+    mentorshipStatus,
+    mentorshipInterest,
+    execInterest,
+    execCoachingStatus,
+    execCoachingInterest,
+  }: EmployeeProfile): FormValues => ({
+    organizationTypeInterest:
+      organizationTypeInterest?.map((x) => x.value) ?? [],
+    moveInterest: moveInterest?.map((x) => x.value) ?? [],
+    mentorshipStatus: mentorshipStatusToFormValues(mentorshipStatus),
+    mentorshipInterest: mentorshipInterest?.map((x) => x.value) ?? [],
+    execInterest: boolToYesNo(execInterest),
+    execCoachingStatus: execCoachingStatusToFormValues(execCoachingStatus),
+    execCoachingInterest: execCoachingInterest?.map((x) => x.value) ?? [],
+  });
```
Could this be defined outside of the component so it doesn't need to be recreated on every mount?

yonikid15 replied:
Moved in [22c8483](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/22c8483a97b8f6cf19dacc81c622f83c80124ce0)

esizer commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,503 @@
+                id="moveInterest"
+                items={moveInterestOptions}
+              />
+              <Separator data-h2-margin="base(0)" />
```
Need the decorative prop here

yonikid15 replied:
Always forget this 😅 [5d09538](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/5d09538675cde5129e1d0480f266af18256951a4)

esizer commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,503 @@
+  }));
+
+  const moveInterestOptions: CheckboxOption[] = unpackMaybes(
+    careerDevelopmentOptions?.moveInterest,
+  ).map(({ value }) => ({
+    value,
+    label: intl.formatMessage(getMoveInterest(value)),
+  }));
+
+  const mentorshipStatusOptions: Radio[] = [
+    { value: MentorshipStatus.NOT_PARTICIPATING },
+    ...unpackMaybes(careerDevelopmentOptions?.mentorship),
+    { value: MentorshipStatus.MENTEE_AND_MENTOR },
+  ].map(({ value }) => ({
+    value,
+    label: intl.formatMessage(getMentorshipStatus(value)),
+  }));
+
+  const mentorshipInterestOptions: Radio[] = unpackMaybes(
+    careerDevelopmentOptions?.mentorship,
+  ).map(({ value }) => ({
+    value,
+    label: intl.formatMessage(getMentorshipInterest(value)),
+  }));
+
+  const execCoachingStatusOptions: Radio[] = [
+    { value: ExecCoachingStatus.NOT_PARTICIPATING },
+    ...unpackMaybes(careerDevelopmentOptions?.execCoaching),
+    { value: ExecCoachingStatus.LEARNING_AND_COACHING },
+  ].map(({ value }) => ({
+    value,
+    label: intl.formatMessage(getExecCoachingStatus(value)),
+  }));
+
+  const execCoachingInterestOptions: Radio[] = unpackMaybes(
+    careerDevelopmentOptions?.execCoaching,
+  ).map(({ value }) => ({
+    value,
+    label: intl.formatMessage(getExecCoachingInterest(value)),
+  }));
```
Seems like we have a lot of loops here for only 3-4 items. Would it be better to just write these arrays directly in the prop?

yonikid15 replied:
Moved in [6f560cb](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/6f560cb4f502bfa0da754e1146a8c2fdb03d2855)

brindasasi commented on apps/web/src/pages/EmployeeProfile/EmployeeProfilePage.tsx:
```diff
@@ -212,7 +226,10 @@ const EmployeeProfilePage = () => {
-        <EmployeeProfile userQuery={data.me} />
+        <EmployeeProfile
+          userQuery={data.me}
+          careerDevelopmentOptionsQuery={data}
```
`data` already has me inside of it. I don't think you need it twice in here.  Remove `userQuery` from the props and you can it from `careerDevelopmentOptionsQuery` inside the `EmployeeProfile` component.

yonikid15 replied:
Hmm, I'm not sure what you mean here? I kept them separate since they are both fragments that will be used in the nested component.

petertgiles replied:
This makes sense.  One is a fragment on `User` and one is a fragment on `Query`, right?

yonikid15 commented:
> * The word counter is working weird.  It doesn't count when I add the word in the next line.
I couldn't recreate it. I might be doing it wrong. Either way I would say this is out of scope.

> * For the larger text it looks like this . Could we wrap it ?
We should fix this across the app. In the <FieldDisplay />.


> * I'm unable to get the missing information screen shot that you have posted above.  As a new government employee I get these sections as below

Isn't this normal behaviour? Its the same as all the other forms? I believe the design you are looking at hasn't been done yet.


> * Minor - Order of the section is different from figma you have posted above . Interest in promotions should come first

Okay! Nice catch!

brindasasi commented:
> > * I'm unable to get the missing information screen shot that you have posted above.  As a new government employee I get these sections as below
> 
> Isn't this normal behaviour? Its the same as all the other forms? I believe the design you are looking at hasn't been done yet.

>  It's in the A/C though.  @petertgiles  do you know we have a plan to build this in a sep story? 

![Screenshot 2025-02-10 at 10 00 55 AM](https://github.com/user-attachments/assets/d3ab2113-7cd8-4854-a21e-e2eba381d01e)

petertgiles commented:
I think we did agree to use our existing inline form here so it might look a little different from the design because of that.

The way our forms work is if all fields are missing then a single message appears in the box instead of listing every missing field, like the language form here.  If at least one field is filled then the missing fields are listed, like the gov employee box.
![image](https://github.com/user-attachments/assets/c148fdf1-65f8-450b-835c-0a11042d7267)

Does this help?

brindasasi commented:
> I think we did agree to use our existing inline form here so it might look a little different from the design because of that.
> 
> The way our forms work is if all fields are missing then a single message appears in the box instead of listing every missing field, like the language form here. If at least one field is filled then the missing fields are listed, like the gov employee box. ![image](https://private-user-images.githubusercontent.com/8978655/411650501-c148fdf1-65f8-450b-835c-0a11042d7267.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMTc4NTQsIm5iZiI6MTczOTIxNzU1NCwicGF0aCI6Ii84OTc4NjU1LzQxMTY1MDUwMS1jMTQ4ZmRmMS02NWY4LTQ1MGItODM1Yy0wYTExMDQyZDcyNjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTBUMTk1OTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWY2YTAxNWU2M2UxNzY2Yzk3YTJjNDliM2Y4MDFmY2E2YTdjNjFjYTc2YTNmNjFjMWRjOTEyYjU2MTAwNDA4YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.9PRYDmj8HXyhTTlIWcHkTLAL8GzEE06V_OYXR4lPXdA)
> 
> Does this help?

I get that . But this PR doesn't have that code for "Missing information" at all and Yoni's understanding is that , it will be done in different PR. 

Currently what's happening , when I very first time load the form I get this form 
![410599498-810985f1-a7fa-4eed-bbc5-7c00ac279cb7](https://github.com/user-attachments/assets/0ea46186-43e6-456b-8939-1ac8a8ff7f44)

When I enter values for all the mandatory fields and leave out the optional fields as is , it displays like this 
![Screenshot 2025-02-10 at 3 04 09 PM](https://github.com/user-attachments/assets/709ed8e3-36e6-41a9-86da-0dfa3f10e61d)

There is no logic here to call the "Missing information" portion on any situation. 
I just want to make sure whether this is acceptable.

yonikid15 commented on apps/web/src/pages/EmployeeProfile/EmployeeProfilePage.tsx:
```diff
@@ -39,15 +47,20 @@ const EmployeeProfile_Fragment = graphql(/** GraphQL */ `
 
 interface EmployeeProfileProps {
   userQuery: FragmentType<typeof EmployeeProfile_Fragment>;
+  careerDevelopmentOptions?: EmployeeProfileCareerDevelopmentOptionsFragment;
```
This is outdated, fixed in [ae85322](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/ae85322229bfc6af815887d60fae74f050e6e635)

yonikid15 commented on apps/web/src/pages/EmployeeProfile/EmployeeProfilePage.tsx:
```diff
@@ -209,10 +226,18 @@ const EmployeeProfilePage = () => {
+  const careerDevelopmentOptions = getFragment(
+    EmployeeProfileCareerDevelopmentOptions_Fragment,
+    data,
+  );
```
This is outdated, fixed in [ae85322](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/ae85322229bfc6af815887d60fae74f050e6e635)

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,509 @@
+    execCoachingStatus,
+    execCoachingInterest,
+  }: FormValues) => {
+    if (!userAuthInfo?.id) {
+      throw new UnauthorizedError();
+    }
```
Nice catch, I was following the format we had for the other form and didn't notice. Changed in [22c8483](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/22c8483a97b8f6cf19dacc81c622f83c80124ce0)

Should I change it in the GoalsWorkStyleForm as well?

petertgiles replied:
Yep, this should never happen.  However, this check provides type narrowing to prove that we're not going to pass null into the mutation.

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,509 @@
+              execCoachingStatus,
+              execCoachingInterest,
+            },
+            { keepDirty: true },
```
That's true! Changed in [6419c95](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/6419c958e4b23b77362519784ae5bc821a2c7c67)

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,509 @@
+                  ].map((value) => ({
+                    value,
+                    label: intl.formatMessage(getExecInterest(value)),
+                  }))}
```
Changed in [8854643](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/8854643ad46681e9498b775d52efee6b84bab9c1)

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,509 @@
+                  rules={{
+                    required: intl.formatMessage(errorMessages.required),
+                  }}
+                />
```
This is outdated, it should have been changed in [896a341](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/896a341aec2eb676e9133256bd9c1cbe69e918c1)

brindasasi commented:
> > There is no logic here to call the "Missing information" portion on any situation which is fine if we agreed to go with that. I just want to make sure whether this is acceptable. Also please remove it from A/C as well.
> 
> @brindasasi The design with the "Missing information" hasn't been done yet, so it should be acceptable to remove it from the A/C.

yeah ..  we had a separate story for that yesterday scrum. Could be removed in A/C here.

petertgiles commented:
There are changes to this form ready to be worked on now:
https://github.com/GCTC-NTGC/gc-digital-talent/issues/12745

I've asked Yoni to finish up this PR with the AC as-is, though perhaps the same level of "doneness" is not necessary here.  We'll immediately begin the changes after this PR is merged.

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/utils.ts:
```diff
@@ -0,0 +1,144 @@
+import { FormValues } from "./CareerDevelopmentSection";
+
+export const mentorshipStatusToFormValues = (
+  mentorshipStatus: EmployeeProfile["mentorshipStatus"],
```
I'm trying to keep it as closely typed to what is returned from the data. In this case:

`EmployeeProfile["mentorshipStatus"] = Maybe<LocalizedMentorship[]>`

Would you prefer I change it to `Maybe<LocalizedMentorship[]>` instead?

petertgiles replied:
I like this type definition - I think it's very expressive about where the type comes from.

esizer commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/Display.tsx:
```diff
@@ -0,0 +1,201 @@
+  getLabels,
+} from "./utils";
+
+interface DisplayProps {
+  employeeProfile: EmployeeProfileCareerDevelopmentFragment;
+  careerDevelopmentOptions: EmployeeProfileCareerDevelopmentOptionsFragment;
+}
```
These should be something closer to `FragmentType<typeof EmployeeProfileCareerDevelopment_Fragment>`

yonikid15 replied:
Refactored in [c3aaaf2](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/c3aaaf22dc15805bb7efdf816ec2ec3e37dc76a1)

esizer commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/Display.tsx:
```diff
@@ -0,0 +1,201 @@
+        )}
+      </ToggleForm.FieldDisplay>
+      <ToggleForm.FieldDisplay
+        hasError={!!execInterest !== execInterest}
```
I could be wrong, but doesn't this always evaluate to the same thing? :sweat_smile:

yonikid15 replied:
It's a little trick I saw recently :laughing:. ExecInterest type can be `execInterest: boolean | null | undefined`. If...
-  !!true === true
- !!false (or false) === false
- !!null (or false)  === null
- !!undefined (or false) === undefined

Since it's only false if it's null or undefined the little trick above works.
I forgot to revert this back to `empty(execInterest)` haha

yonikid15 replied:
Fixed in [81b268c](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/81b268cfbfc291a2819e389ed49ff11c60235864)

esizer commented on apps/web/src/pages/EmployeeProfile/components/GoalsWorkStyleSection/GoalsWorkStyleSection.tsx:
```diff
@@ -123,12 +122,8 @@ const GoalsWorkStyleSection = ({
-
     return executeMutation({
-      id: userAuthInfo?.id,
+      id: userId,
```
Is there a reason we switched from just reading the user ID from the already queried auth data to just drilling it down?

yonikid15 replied:
Short answer is typescript wants a string not a string | undefined. That's why someone added the check earlier.

```ts   
if (!userAuthInfo?.id) {
   throw new UnauthorizedError();
}
```

I can't remember if there's another way we were doing this?

esizer replied:
You can just throw an error to prevent it and create a type guard

petertgiles replied:
Thanks for switching back to the useAuth hook and using the `if` to narrow the type.

yonikid15 commented on apps/web/src/pages/EmployeeProfile/components/CareerDevelopmentSection/CareerDevelopmentSection.tsx:
```diff
@@ -0,0 +1,509 @@
+    ),
+    execCoachingInterest:
+      initialData.execCoachingInterest?.map((x) => x.value) ?? [],
+  });
```
This is moved in [22c8483](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/22c8483a97b8f6cf19dacc81c622f83c80124ce0)

yonikid15 commented:
@brindasasi 
- Fixed the text overflowing the page [787d7ff](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/787d7ff8b1b3223ca91d209a799dc2f4ebc5aa73)
- Fixed the order of the form fields [50e162c](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/50e162c5a3705ee4b70ebebbd37c7467afdd58ca)
- I wasn't able to recreate the word counter issue you found. However, that's not part of this issue haha. We should make a new issue to tackle that.

Anything else I might of missed?

brindasasi commented:
Will review in a moment ! Thanks !

brindasasi commented on apps/web/src/pages/EmployeeProfile/EmployeeProfilePage.tsx:
```diff
@@ -39,15 +48,24 @@ const EmployeeProfile_Fragment = graphql(/** GraphQL */ `
+  userId,
+  userQuery,
+  careerDevelopmentOptionsQuery,
+}: EmployeeProfileProps) => {
```
Its a small clean up. 
since userId , userQuery are retrieved from data , they can be assigned inside EmployeeProfile component. 
passing the data that you already have causes duplication. 
I meant this way , 

```
const EmployeeProfile = ({
careerDevelopmentOptionsQuery,
}: EmployeeProfileProps) => { 
const intl = useIntl();
 const paths = useRoutes();
 const userQuery = careerDevelopmentOptionsQuery.me;
 const userId = userQuery.id;
 
const user = getFragment(EmployeeProfile_Fragment, userQuery);
```
	
 ...

yonikid15 replied:
I removed the userId, but for the other two props, I think I should leave them since they are both fragments.

petertgiles replied:
This makes sense. One is a fragment on User and one is a fragment on Query, right?

yonikid15 replied:
I could pass down the entire query in the props then split it up there, but it makes it easier with the typing to do it like this.

yonikid15 replied:
Nvm Brinda was right, I couldn't figure out this type error, but Brinda helped me see the light :bulb: 
Changed in [c0ec84c](https://github.com/GCTC-NTGC/gc-digital-talent/pull/12680/commits/c0ec84c3f3e0afadb3e0556d80d7fd2811fa26af)

brindasasi reviewed (approved):
Looks great! Thank you for making the changes. I created sep issue for word counter component as that is out of scope.  Good to go! 🥳👏
